Remove Coq 8.13 from Makefile

It seems VST no longer supports Coq 8.13 (or any least Coq 8.13.0). So we should remove it from Makefile.
https://github.com/PrincetonUniversity/VST/blob/master/Makefile
I tested on Coq 8.13.0 and it reports
```
File ".\msl/sepalg.v", line 160, characters 0-97:
Error: export attribute not supported here

make: *** [Makefile:660: msl/sepalg.vo] Error 1
```

--- a/.github/workflows/coq-action.yml
+++ b/.github/workflows/coq-action.yml
@@ -5,7 +5,6 @@ on:
   push:
     branches:
       - master
-      - tcarstens/conclib
   pull_request:
     branches:
       - '**'

--- a/.github/workflows/coq-action.yml
+++ b/.github/workflows/coq-action.yml
@@ -5,9 +5,9 @@ on:
   push:
     branches:
       - master
-#  pull_request:
-#    branches:
-#      - '**'
+  pull_request:
+    branches:
+      - '**'
 
 jobs:
   build:

--- a/.github/workflows/coq-action.yml
+++ b/.github/workflows/coq-action.yml
@@ -5,9 +5,9 @@ on:
   push:
     branches:
       - master
-  pull_request:
-    branches:
-      - '**'
+#  pull_request:
+#    branches:
+#      - '**'
 
 jobs:
   build:

--- a/.github/workflows/coq-action.yml
+++ b/.github/workflows/coq-action.yml
@@ -84,6 +84,7 @@ jobs:
           - '8.14'
           - '8.15'
         make_target:
+          - assumptions.txt
           - test
           - test2
           - test3

--- a/.github/workflows/tcb-action.yml
+++ b/.github/workflows/tcb-action.yml
@@ -17,4 +17,4 @@ jobs:
   run:
     steps:
       - name: 'Fail because TCB is modified'
-	run: echo 'TCB modified -- automatic failure' ; exit 1
+        run: echo 'TCB modified -- automatic failure' ; exit 1

--- a/.github/workflows/tcb-action.yml
+++ b/.github/workflows/tcb-action.yml
@@ -15,6 +15,7 @@ on:
 
 jobs:
   run:
+    runs-on: ubuntu-latest
     steps:
       - name: 'Fail because TCB is modified'
         run: echo 'TCB modified -- automatic failure' ; exit 1

--- a/.github/workflows/tcb-action.yml
+++ b/.github/workflows/tcb-action.yml
@@ -16,6 +16,5 @@ on:
 jobs:
   run:
     steps:
-      - name: "Fail because TCB is modified"
-        id: tcb_fail
-	run: echo "TCB modified -- automatic failure" ; exit 1
+      - name: 'Fail because TCB is modified'
+	run: echo 'TCB modified -- automatic failure' ; exit 1

--- a/.github/workflows/tcb-action.yml
+++ b/.github/workflows/tcb-action.yml
@@ -1,0 +1,21 @@
+---
+name: TCB
+
+on:
+  push:
+    branches:
+      - master
+    paths:
+      - veric/NullExtension.v
+  pull_request:
+    branches:
+      - '**'
+    paths:
+      - veric/NullExtension.v
+
+jobs:
+  run:
+    steps:
+      - name: "Fail because TCB is modified"
+        id: tcb-fail
+	run: echo "TCB modified -- automatic failure" ; exit 1

--- a/.github/workflows/tcb-action.yml
+++ b/.github/workflows/tcb-action.yml
@@ -17,5 +17,5 @@ jobs:
   run:
     steps:
       - name: "Fail because TCB is modified"
-        id: tcb-fail
+        id: tcb_fail
 	run: echo "TCB modified -- automatic failure" ; exit 1

--- a/.github/workflows/tcb-action.yml
+++ b/.github/workflows/tcb-action.yml
@@ -6,12 +6,12 @@ on:
     branches:
       - master
     paths:
-      - veric/NullExtension.v
+      - veric/tcb.v
   pull_request:
     branches:
       - '**'
     paths:
-      - veric/NullExtension.v
+      - veric/tcb.v
 
 jobs:
   run:

--- a/Makefile
+++ b/Makefile
@@ -21,7 +21,7 @@ COQLIB=$(shell $(COQC) -where | tr -d '\r' | tr '\\' '/')
 
 # Check Coq version
 
-COQVERSION= 8.13.0 or-else 8.13.1 or-else 8.13.2 or-else 8.14.0 or-else 8.14.1 or-else 8.15.0 or-else 8.15.1
+COQVERSION= 8.14.0 or-else 8.14.1 or-else 8.15.0 or-else 8.15.1 or-else 8.15.2
 
 COQV=$(shell $(COQC) -v)
 ifneq ($(IGNORECOQVERSION),true)
@@ -440,7 +440,7 @@ VERIC_FILES= \
   juicy_mem.v juicy_mem_lemmas.v local.v juicy_mem_ops.v juicy_safety.v juicy_extspec.v \
   semax.v semax_lemmas.v semax_conseq.v semax_call.v semax_straight.v semax_loop.v semax_switch.v \
   initial_world.v Clight_initial_world.v initialize.v semax_prog.v semax_ext.v SeparationLogic.v SeparationLogicSoundness.v  \
-  NullExtension.v SequentialClight.v superprecise.v jstep.v address_conflict.v valid_pointer.v coqlib4.v \
+  NullExtension.v SequentialClight.v SequentialClight2.v superprecise.v jstep.v address_conflict.v valid_pointer.v coqlib4.v \
   semax_ext_oracle.v mem_lessdef.v Clight_mem_lessdef.v age_to_resource_at.v aging_lemmas.v Clight_aging_lemmas.v ghost_PCM.v mpred.v ghosts.v invariants.v
 
 ZLIST_FILES= \

--- a/Makefile
+++ b/Makefile
@@ -440,7 +440,7 @@ VERIC_FILES= \
   juicy_mem.v juicy_mem_lemmas.v local.v juicy_mem_ops.v juicy_safety.v juicy_extspec.v \
   semax.v semax_lemmas.v semax_conseq.v semax_call.v semax_straight.v semax_loop.v semax_switch.v \
   initial_world.v Clight_initial_world.v initialize.v semax_prog.v semax_ext.v SeparationLogic.v SeparationLogicSoundness.v  \
-  NullExtension.v SequentialClight.v SequentialClight2.v superprecise.v jstep.v address_conflict.v valid_pointer.v coqlib4.v \
+  NullExtension.v SequentialClight.v SequentialClight2.v tcb.v superprecise.v jstep.v address_conflict.v valid_pointer.v coqlib4.v \
   semax_ext_oracle.v mem_lessdef.v Clight_mem_lessdef.v age_to_resource_at.v aging_lemmas.v Clight_aging_lemmas.v ghost_PCM.v mpred.v ghosts.v invariants.v
 
 ZLIST_FILES= \
@@ -878,6 +878,10 @@ progs64: _CoqProject  $(PROGS64_FILES:%.v=progs64/%.vo)
 
 VSUpile: floyd/proofauto.vo floyd/library.vo floyd/VSU.vo
 	cd progs/VSUpile; $(MAKE) BUNDLED=true
+
+assumptions.txt: veric/tcb.vo
+	$(COQC) $(COQFLAGS) veric/tcb.v > assumptions.txt
+	bash util/check_assumptions.sh
 
 # $(CC_TARGET): compcert/make
 #	(cd compcert; ./make)

--- a/util/check_assumptions.sh
+++ b/util/check_assumptions.sh
@@ -1,20 +1,21 @@
 #!/usr/bin/env bash
-grep -v -f - assumptions.txt <<EOF 
+grep -E -v -f - assumptions.txt <<EOF 
 ^Axioms:
-^ClassicalDedekindReals.sig_not_dec[^[:alnum:]_]
-^ClassicalDedekindReals.sig_forall_dec[^[:alnum:]_]
-^prop_ext[^[:alnum:]_]
-^Events.inline_assembly_sem[^[:alnum:]_]
-^functional_extensionality_dep[^[:alnum:]_]
-^Events.external_functions_sem[^[:alnum:]_]
-^Eqdep.Eq_rect_eq.eq_rect_eq[^[:alnum:]_]
-^Classical_Prop.classic[^[:alnum:]_]
+^ClassicalDedekindReals.sig_not_dec( |[[:space:]]*$)
+^ClassicalDedekindReals.sig_forall_dec( |[[:space:]]*$)
+^prop_ext( |[[:space:]]*$)
+^Events.inline_assembly_sem( |[[:space:]]*$)
+^functional_extensionality_dep( |[[:space:]]*$)
+^Events.external_functions_sem( |[[:space:]]*$)
+^Eqdep.Eq_rect_eq.eq_rect_eq( |[[:space:]]*$)
+^Classical_Prop.classic( |[[:space:]]*$)
 ^[ ]
 EOF
 if [ $? -ne 0 ]; then
    echo "Success! No unexpected assumptions"
 else
     echo "Failure: Unexpected assumptions"
+    
     exit 1
 fi
 

--- a/util/check_assumptions.sh
+++ b/util/check_assumptions.sh
@@ -1,0 +1,20 @@
+#!/usr/bin/env bash
+grep -v -f - assumptions.txt <<EOF 
+^Axioms:
+^ClassicalDedekindReals.sig_not_dec[^[:alnum:]_]
+^ClassicalDedekindReals.sig_forall_dec[^[:alnum:]_]
+^prop_ext[^[:alnum:]_]
+^Events.inline_assembly_sem[^[:alnum:]_]
+^functional_extensionality_dep[^[:alnum:]_]
+^Events.external_functions_sem[^[:alnum:]_]
+^Eqdep.Eq_rect_eq.eq_rect_eq[^[:alnum:]_]
+^Classical_Prop.classic[^[:alnum:]_]
+^[ ]
+EOF
+if [ $? -ne 0 ]; then
+   echo "Success! No unexpected assumptions"
+else
+    echo "Failure: Unexpected assumptions"
+    exit 1
+fi
+

--- a/veric/NullExtension.v
+++ b/veric/NullExtension.v
@@ -1,13 +1,13 @@
 Require Import VST.sepcomp.extspec.
 Require Import VST.sepcomp.step_lemmas.
 Require Import VST.veric.base.
-(*Lenb - do we need these?
-  Require Import VST.veric.Clight_core.
-Require Import VST.veric.Clight_lemmas.*)
 Require Import VST.veric.juicy_extspec.
 Require Import VST.veric.juicy_mem.
+Require Import VST.veric.compspecs.
+Require Import VST.veric.semax_prog.
+Require Import VST.veric.SequentialClight2.
 
-Definition dryspec : external_specification juicy_mem external_function unit
+Definition juicyspec : external_specification juicy_mem external_function unit
   := Build_external_specification juicy_mem external_function unit
      (*ext_spec_type*)
      (fun ef => False)
@@ -16,16 +16,89 @@ Definition dryspec : external_specification juicy_mem external_function unit
      (*ext_spec_post*)
      (fun ef Hef ge ty vl m z => False)
      (*ext_spec_exit*)
-     (fun rv m z => False).
+     (fun rv m z => True).
 
 Definition Espec : OracleKind.
- refine (Build_OracleKind unit (Build_juicy_ext_spec _ dryspec _ _ _ _ _ _)).
+ refine (Build_OracleKind unit (Build_juicy_ext_spec _ juicyspec _ _ _ _ _ _)).
 Proof.
 simpl; intros; contradiction.
 simpl; intros; contradiction.
 simpl; intros; intros ? ? ? ?; contradiction.
 simpl; intros; contradiction.
-simpl; intros; intros ? ? ? ?; contradiction.
-simpl; intros; intros ? ? ? ?; contradiction.
+repeat intro; auto.
+repeat intro; auto.  
 Defined.
 
+Definition dryspec : external_specification mem external_function unit
+  := Build_external_specification mem external_function unit
+     (*ext_spec_type*)
+     (fun ef => False)
+     (*ext_spec_pre*)
+     (fun ef Hef ge tys vl m z => False)
+     (*ext_spec_post*)
+     (fun ef Hef ge ty vl m z => False)
+     (*ext_spec_exit*)
+     (fun rv m z => True).
+
+Lemma NullExtension_whole_program_sequential_safety:
+   forall {CS: compspecs}
+     (prog: Clight.program) V G m,
+     @semax_prog Espec CS prog tt V G ->
+     Genv.init_mem prog = Some m ->
+     exists b, exists q, exists m',
+       Genv.find_symbol (Genv.globalenv prog) (prog_main prog) = Some b /\
+       semantics.initial_core  (Clight_core.cl_core_sem (Clight.globalenv prog))
+           0 m q m' (Vptr b Ptrofs.zero) nil /\
+       forall n,
+        @dry_safeN _ _ _ unit (semax.genv_symb_injective)
+          (Clight_core.cl_core_sem (Clight.globalenv prog)) dryspec
+           (Clight.genv_genv 
+            (Clight.Build_genv (Genv.globalenv prog) (Ctypes.prog_comp_env prog)) )
+             n tt q m'.
+Proof.
+intros.
+assert (dessicate : forall ef : external_function,
+                   juicy_mem ->
+                   @ext_spec_type juicy_mem external_function
+                     (@OK_ty NullExtension.Espec) (@OK_spec NullExtension.Espec) ef ->
+                   @ext_spec_type mem external_function 
+                     (@OK_ty NullExtension.Espec) dryspec ef). {
+  intros. assumption.
+}
+apply (@whole_program_sequential_safety CS NullExtension.Espec
+ tt dryspec dessicate) with (V:=V) (G:=G); auto.
+-
+split; intros; try assumption.
+split; intros; try assumption.
+split; intros; try assumption.
+-
+hnf; intros; contradiction.
+-
+repeat intro; auto.
+hnf. auto.
+Qed.
+
+Lemma module_sequential_safety : (*TODO*)
+   forall {CS: compspecs} (prog: Clight.program) (V: mpred.varspecs) 
+      (G: mpred.funspecs) ora m f f_id f_b f_body args,
+     let ge := Genv.globalenv prog in
+     let ext_link := SeparationLogic.ext_link_prog prog in
+     let spec := SeparationLogic.add_funspecs NullExtension.Espec ext_link G in
+     let tys := sig_args (ef_sig f) in
+     let rty := sig_res (ef_sig f) in
+     let sem := juicy_core_sem (Clight_core.cl_core_sem (Clight.Build_genv ge (prog_comp_env prog))) in
+     @semax_prog spec CS prog ora V G ->
+     fun_id ext_link f = Some f_id ->
+     Genv.find_symbol ge f_id = Some f_b ->
+     Genv.find_funct  ge (Vptr f_b Ptrofs.zero) = Some f_body ->
+     forall x : ext_spec_type (@OK_spec spec) f,
+     ext_spec_pre (@OK_spec spec) f x (semax.genv_symb_injective ge) tys args ora m ->
+     exists q,
+       semantics.initial_core sem 
+         0 (*additional temporary argument - TODO (Santiago): FIXME*)
+             m q m
+              (Vptr f_b Ptrofs.zero) args /\
+       forall n, safeN_(genv_symb := @semax.genv_symb_injective _ _)(Hrel := juicy_extspec.Hrel)
+            sem  (upd_exit (@OK_spec spec) x (semax.genv_symb_injective ge)) 
+           ge n ora q m.
+Abort.

--- a/veric/NullExtension.v
+++ b/veric/NullExtension.v
@@ -29,16 +29,7 @@ repeat intro; auto.
 repeat intro; auto.  
 Defined.
 
-Definition dryspec : external_specification mem external_function unit
-  := Build_external_specification mem external_function unit
-     (*ext_spec_type*)
-     (fun ef => False)
-     (*ext_spec_pre*)
-     (fun ef Hef ge tys vl m z => False)
-     (*ext_spec_post*)
-     (fun ef Hef ge ty vl m z => False)
-     (*ext_spec_exit*)
-     (fun rv m z => True).
+Definition dryspec := juicy_dry_ext_spec_make _ juicyspec.
 
 Lemma NullExtension_whole_program_sequential_safety:
    forall {CS: compspecs}
@@ -68,9 +59,10 @@ assert (dessicate : forall ef : external_function,
 apply (@whole_program_sequential_safety CS NullExtension.Espec
  tt dryspec dessicate) with (V:=V) (G:=G); auto.
 -
+split; intros; try assumption; try contradiction.
 split; intros; try assumption.
-split; intros; try assumption.
-split; intros; try assumption.
+split; repeat intro; auto.
+split; repeat intro; auto.
 -
 hnf; intros; contradiction.
 -

--- a/veric/SequentialClight.v
+++ b/veric/SequentialClight.v
@@ -8,7 +8,6 @@ Require Import VST.sepcomp.event_semantics.
 Require Import VST.veric.SeparationLogic.
 Require Import VST.veric.juicy_extspec.
 Require Import VST.veric.juicy_mem.
-Require VST.veric.NullExtension.
 Require Import VST.veric.SeparationLogicSoundness.
 Require Import VST.sepcomp.extspec.
 Require Import VST.msl.msl_standard.

--- a/veric/SequentialClight2.v
+++ b/veric/SequentialClight2.v
@@ -8,8 +8,7 @@ Require Import VST.sepcomp.event_semantics.
 Require Import VST.veric.SeparationLogic.
 Require Import VST.veric.juicy_extspec.
 Require Import VST.veric.juicy_mem.
-Require VST.veric.NullExtension.
-(*Require Import VST.veric.Clight_sim.*)
+(*Require VST.veric.NullExtension. *)
 Require Import VST.veric.SeparationLogicSoundness.
 Require Import VST.sepcomp.extspec.
 Require Import VST.msl.msl_standard.
@@ -384,26 +383,25 @@ Qed.
                ext_spec_type dryspec ef)
      (JDE: juicy_dry_ext_spec _ (@JE_spec OK_ty OK_spec) dryspec dessicate)
      (DME: ext_spec_mem_evolve _ dryspec)
-     prog V G m,
-     @semax_prog Espec (*NullExtension.Espec*) CS prog V G ->
+     (PAE: semax_prog.postcondition_allows_exit Espec tint)
+     (prog: Clight.program) V G m,
+     @semax_prog Espec CS prog initial_oracle V G ->
      Genv.init_mem prog = Some m ->
      exists b, exists q, exists m',
        Genv.find_symbol (Genv.globalenv prog) (prog_main prog) = Some b /\
        initial_core  (cl_core_sem (globalenv prog))
            0 m q m' (Vptr b Ptrofs.zero) nil /\
        forall n,
-        @dry_safeN _ _ _ OK_ty (genv_symb_injective)
-            (coresem_extract_cenv  (cl_core_sem (globalenv prog))
-                (prog_comp_env prog)) 
-            (*(dryspec  OK_ty)*) dryspec
+        @dry_safeN _ _ _ OK_ty (semax.genv_symb_injective)
+          (cl_core_sem (globalenv prog)) dryspec
             (Build_genv (Genv.globalenv prog) (prog_comp_env prog)) 
              n initial_oracle q m'.
 Proof.
  intros.
- destruct (@semax_prog_rule' (*NullExtension.*)Espec CS _ _ _ _ 
+ destruct (@semax_prog_rule Espec CS _ _ _ _ 
      0 (*additional temporary argument - TODO (Santiago): FIXME*)
-     H H0) as [b [q [[H1 H2] H3]]].
- destruct (H3 O initial_oracle) as [jmx [H4x [H5x [H6x [H7x _]]]]].
+     initial_oracle PAE H H0) as [b [q [[H1 H2] H3]]].
+ destruct (H3 O) as [jmx [H4x [H5x [H6x [H7x _]]]]].
  destruct (H2 jmx H4x) as [jmx' [H8x H8y]].
  exists b, q, (m_dry jmx').
  split3; auto.
@@ -411,7 +409,7 @@ Proof.
  subst. simpl. clear H5x H6x H7x H8y.
  forget (m_dry jmx) as m. clear jmx.
  intro n.
- specialize (H3 n initial_oracle).
+ specialize (H3 n).
  destruct H3 as [jm [? [? [? [? _]]]]].
  unfold semax.jsafeN in H6.
  subst m.
@@ -445,12 +443,7 @@ Proof.
      apply IHn; auto. lia.
      replace (level m'') with n0 by lia. auto.
  -
-(*
-   assert (JDE1': ext_spec_type dryspec = ext_spec_type OK_spec)
-      by apply JDE.
-*)
-(*   destruct JDE as [JDE1 [JDE2 [JDE3 JDE4]]]. *)
-   destruct dryspec as [ty pre post exit]. simpl in *. (* subst ty. *)
+   destruct dryspec as [ty pre post exit]. simpl in *.
    destruct JE_spec as [ty' pre' post' exit']. simpl in *.
    change (level (m_phi jm)) with (level jm) in *.
    destruct JDE as [JDE1 [JDE2 JDE3]].
@@ -537,35 +530,9 @@ Proof.
  - eapply safeN_halted; eauto.
     apply JDE. auto.
  Unshelve. simpl. split; [apply Share.nontrivial | hnf]. exists None; constructor.
-all: fail.
 Qed.
 
 Require Import VST.veric.juicy_safety.
 
 Definition fun_id (ext_link: Strings.String.string -> ident) (ef: external_function) : option ident :=
   match ef with EF_external id sig => Some (ext_link id) | _ => None end.
-
-Print genv.
-
-Lemma module_sequential_safety : (*TODO*)
-   forall {CS: compspecs} (prog: program) (V: varspecs) (G: funspecs) ora m f f_id f_b f_body args,
-     let ge := Genv.globalenv prog in
-     let ext_link := ext_link_prog prog in
-     let spec := add_funspecs NullExtension.Espec ext_link G in
-     let tys := sig_args (ef_sig f) in
-     let rty := sig_res (ef_sig f) in
-     let sem := juicy_core_sem (cl_core_sem (Build_genv ge (prog_comp_env prog))) in
-     @semax_prog spec CS prog V G ->
-     fun_id ext_link f = Some f_id ->
-     Genv.find_symbol ge f_id = Some f_b ->
-     Genv.find_funct  ge (Vptr f_b Ptrofs.zero) = Some f_body ->
-     forall x : ext_spec_type (@OK_spec spec) f,
-     ext_spec_pre (@OK_spec spec) f x (genv_symb_injective ge) tys args ora m ->
-     exists q,
-       initial_core sem 
-         0 (*additional temporary argument - TODO (Santiago): FIXME*)
-             m q m
-              (Vptr f_b Ptrofs.zero) args /\
-       forall n, safeN_(genv_symb := @genv_symb_injective _ _)(Hrel := juicy_extspec.Hrel) (coresem_extract_cenv sem (prog_comp_env prog))
-(upd_exit (@OK_spec spec) x (genv_symb_injective ge)) ge n ora q m.
-Abort.

--- a/veric/juicy_extspec.v
+++ b/veric/juicy_extspec.v
@@ -133,11 +133,15 @@ Section upd_exit.
   Definition upd_exit'' (ef : external_function) (x : ext_spec_type spec ef) ge :=
     upd_exit' (ext_spec_post spec ef x ge (sig_res (ef_sig ef))).
 
-  Program Definition upd_exit {ef : external_function} (x : ext_spec_type spec ef) ge :=
-    Build_juicy_ext_spec _ (upd_exit'' _ x ge) _ _ _.
+  Program Definition upd_exit {ef : external_function} (x : ext_spec_type spec ef) ge
+   : juicy_ext_spec Z :=
+    Build_juicy_ext_spec _ (upd_exit'' _ x ge) _ _ _ _ _ _.
   Next Obligation. intros. eapply JE_pre_hered; eauto. Qed.
   Next Obligation. intros. eapply JE_pre_ext; eauto. Qed.
   Next Obligation. intros. eapply JE_post_hered; eauto. Qed.
+  Next Obligation. intros. eapply JE_post_ext; eauto. Qed.
+  Next Obligation. intros. eapply JE_post_hered; eauto. Qed. 
+  Next Obligation. intros. eapply JE_post_ext; eauto. Qed.
 End upd_exit.
 
 Obligation Tactic := Tactics.program_simpl.

--- a/veric/tcb.v
+++ b/veric/tcb.v
@@ -20,8 +20,8 @@ Definition null_extension_juicyspec : external_specification juicy_mem external_
      (fun rv m z => True).
 
 Theorem VST_sound: 
-  exists Espec : OracleKind,
-  JMeq.JMeq (JE_spec _ (@OK_spec Espec)) null_extension_juicyspec /\
+  {Espec : OracleKind 
+  | JMeq.JMeq (JE_spec _ (@OK_spec Espec)) null_extension_juicyspec /\
   let dryspec :=  juicy_dry_ext_spec_make _ null_extension_juicyspec in 
   forall (CS: compspecs)
      (prog: Clight.program) (initial_oracle: OK_ty)
@@ -37,7 +37,7 @@ Theorem VST_sound:
           (Clight_core.cl_core_sem (Clight.globalenv prog)) dryspec
            (Clight.genv_genv 
             (Clight.Build_genv (Genv.globalenv prog) (Ctypes.prog_comp_env prog)) )
-             n tt q m'.
+             n tt q m'}.
 Proof.
 intros.
 exists NullExtension.Espec.

--- a/veric/tcb.v
+++ b/veric/tcb.v
@@ -1,0 +1,51 @@
+Require Import VST.sepcomp.extspec.
+Require Import VST.sepcomp.step_lemmas.
+Require Import VST.veric.base.
+Require Import VST.veric.juicy_extspec.
+Require Import VST.veric.juicy_mem.
+Require Import VST.veric.compspecs.
+Require Import VST.veric.semax_prog.
+Require Import VST.veric.SequentialClight2.
+Require Import VST.veric.NullExtension.
+
+Definition null_extension_juicyspec : external_specification juicy_mem external_function unit
+  := Build_external_specification juicy_mem external_function unit
+     (*ext_spec_type*)
+     (fun ef => False)
+     (*ext_spec_pre*)
+     (fun ef Hef ge tys vl m z => False)
+     (*ext_spec_post*)
+     (fun ef Hef ge ty vl m z => False)
+     (*ext_spec_exit*)
+     (fun rv m z => True).
+
+Theorem VST_sound: 
+  exists Espec : OracleKind,
+  JMeq.JMeq (JE_spec _ (@OK_spec Espec)) null_extension_juicyspec /\
+  let dryspec :=  juicy_dry_ext_spec_make _ null_extension_juicyspec in 
+  forall (CS: compspecs)
+     (prog: Clight.program) (initial_oracle: OK_ty)
+     (V : mpred.varspecs) (G : mpred.funspecs) (m: mem),
+     @semax_prog Espec CS prog initial_oracle V G ->
+     Genv.init_mem prog = Some m ->
+     exists b, exists q, exists m',
+       Genv.find_symbol (Genv.globalenv prog) (prog_main prog) = Some b /\
+       semantics.initial_core  (Clight_core.cl_core_sem (Clight.globalenv prog))
+           0 m q m' (Vptr b Ptrofs.zero) nil /\
+       forall n,
+        @dry_safeN _ _ _ unit (semax.genv_symb_injective)
+          (Clight_core.cl_core_sem (Clight.globalenv prog)) dryspec
+           (Clight.genv_genv 
+            (Clight.Build_genv (Genv.globalenv prog) (Ctypes.prog_comp_env prog)) )
+             n tt q m'.
+Proof.
+intros.
+exists NullExtension.Espec.
+split.
+reflexivity.
+intros.
+destruct initial_oracle.
+eapply NullExtension_whole_program_sequential_safety; eassumption.
+Qed.
+
+Print Assumptions VST_sound.

